### ISSUE TITLE
BUG in QF algorithms

### ISSUE THREAD

**Issue by nishihatapalmer** [smart-tool/smart#7]
All of the QF algorithms from QF3x and above have a bug.  They do not & the D parameter with the B[] in the inner loop.  The QF2x algorithms are correct.

This doesn't stop the algorithms working, it just reduces their performance, as it will not maintain the phase of the qgrams with each other.  Anding this value means that subsequent qgrams have to have at least one phase the same as the preceding qgrams.

```
if(i >= j) {
	ch = y[i+2];
	ch = (ch<<S) + y[i+1];
	ch = (ch<<S) + y[i];
	D = B[ch & mask];
	if(D == 0) continue;
	else goto more;
} 
```
should read:
```
if(i >= j) {
	ch = y[i+2];
	ch = (ch<<S) + y[i+1];
	ch = (ch<<S) + y[i];
	D &= B[ch & mask];
	if(D == 0) continue;
	else goto more;
} 
```

**Comment by nishihatapalmer** [comment]
This issue is fixed by #10

**Comment by nishihatapalmer** [comment]
Reading the original paper, it does say that `m > Q` must hold.  So `m=Q` is ruled out by the original spec.

The algorithm is still capable of processing patterns of length Q (although it will not achieve good results doing so).

In any case, this is not a bug - this is how the algorithm was designed.

EDIT: oops - responding here to a completely different issue with QF.  Ignore this comment.
